conjur/init_logic.py

```diff
@@ -0,0 +1,120 @@
+            if port is None:
+                port = 443
+
+            ctx = SSL.Context(method=SSL.TLSv1_2_METHOD)
+            # We set the following to false because we want to fetch the
+            # certificates and not use them to validate yet
+            ctx.check_hostname = False
+            ctx.verify_mode = False
+
+            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            sock = SSL.Connection(context=ctx, socket=sock)
+            sock.connect((hostname, port))
+            sock.do_handshake()
+        except Exception:
+            raise Exception(f"Unable to retrieve certificate from {hostname}:{port}")
+
+        logging.info("TLS Connection established. " \
+                     "Fetching certificate from server...")
+
+        return sock
+
+    # pylint: disable=unused-variable
+    @staticmethod
+    def get_certificate(hostname, port, sock, dump_certificate_func):
+        """
+        Method for connecting to Conjur to fetch the certificate chain
+        """
+        try:
+            chain = sock.get_peer_cert_chain()
+            cert = chain[0]
+            fingerprint = cert.digest("sha1").decode("utf-8")
+            # Get the peer chair and format to make it easier to later write to a file
+            readable_certificate=''
+            for (idx, cert) in enumerate(chain):
+                readable_certificate+=str(dump_certificate_func(FILETYPE_PEM, cert),  "utf-8")
+            logging.info("Certificates were fetched successfully")
+        except Exception:
+            raise Exception(f"Unable to retrieve certificate from {hostname}:{port}")
+
+        return fingerprint, readable_certificate
+
+    def write_certificate_to_file(self, fetched_certificate, cert_file_path, force_overwrite_flag):
+        """
+        Method for writing certificate details to a file on the user's machine
+        """
+        if fetched_certificate is not None:
+            self._check_file_exists(cert_file_path, force_overwrite_flag)
+            with open(cert_file_path, 'w') as config_fp:
+                config_fp.write(fetched_certificate)
+
+            sys.stdout.write(f"Wrote certificate to {cert_file_path}\n")
+
+    def write_conjurrc(self, conjurrc_file_path, input_obj, force_overwrite_flag):
+        """
+        Method for writing the conjurrc configuration
+        details needed to create a connection to Conjur
+        """
+        self._check_file_exists(conjurrc_file_path, force_overwrite_flag)
+
+        with open(conjurrc_file_path, 'w') as config_fp:
+            _pretty_print_object = {}
+
+            # Ensure no None fields are saved in conjurrc
+            for attr,value in input_obj.__dict__.items():
+                if value is not None:
+                    _pretty_print_object[str(attr)]=value
+
+            config_fp.write("---\n")
+            yaml.dump(_pretty_print_object, config_fp)
+
+        sys.stdout.write(f"Wrote configuration to {conjurrc_file_path}\n")
```

Comment:
Mixing logging and sys write in unadvisable.

Reply:
I need a log to stdout and a logs that appear when the user runs the command with `-d` which are very different formats. If I allow these logger logs to be printed to the screen they will look like 
```
2020-12-16 17:50:11,921 INFO: ... blah
```
which isn't nice for a user to get back especially when we declare for example that a variable was set. In this example,
I would want `Variable set successfully` and not `2020-12-16 17:50:11,921 INFO: Variable set successfully`

If I remove the formatter, I still get `INFO:root blah`. Is there a way to overcome this so I can choose one way of writing to stdout?

Reply:
I need a log to stdout and a logs that appear when the user runs the command with `-d` which are very different formats. If I allow these logger logs to be printed to the screen they will look like 
```
2020-12-16 17:50:11,921 INFO: ... blah
```
which isn't nice for a user to get back especially when we declare for example that a variable was set. In this example,
I would want `Variable set successfully` and not `2020-12-16 17:50:11,921 INFO: Variable set successfully`

I dont believe I can have more than one formatter. Is there a way to overcome this so I can choose one way of writing to stdout?

Reply:
You can expands the logging framework to include a method that will do this for you. Ideally, you would want to use the same logger object but chose if you want the full output or just the raw string.

Reply:
The separation will not be clear in code what is written in stdout and which is in log. IMO Stdout should be the communication between cli and user. Logger should be reserved for the inner workings of the app

Reply:
Fair enough but we're hard-coding the exact file descriptor that we are using each time. Ideally this could be configurable as then you can allow the capability to write the output to a file easy. It's not a blocker but it's something that will make future expansion much, much easier.